File: backend/src/app/schemas/profiles.py
"""Pydantic schemas for profiles."""

from typing import Any

from pydantic import BaseModel, Field


class ProfileBase(BaseModel):
    """Base profile schema."""

    theme: str = Field(default="system")
    color_scheme: str | None = None
    font_size: str = Field(default="medium")
    contrast: str = Field(default="normal")
    dashboard_layout: dict[str, Any] | None = None
    favorite_devices: list[int] | None = None
    default_room: str | None = None
    show_weather: bool = True
    show_clock: bool = True
    reduce_motion: bool = False
    high_contrast: bool = False
    large_touch_targets: bool = False
    kiosk_auto_dim: bool = True
    kiosk_dim_start: str | None = None
    kiosk_dim_end: str | None = None
    kiosk_dim_brightness: int = Field(default=30, ge=0, le=100)


class ProfileUpdate(BaseModel):
    """Schema for updating a profile."""

    theme: str | None = None
    color_scheme: str | None = None
    font_size: str | None = None
    contrast: str | None = None
    dashboard_layout: dict[str, Any] | None = None
    favorite_devices: list[int] | None = None
    default_room: str | None = None
    show_weather: bool | None = None
    show_clock: bool | None = None
    reduce_motion: bool | None = None
    high_contrast: bool | None = None
    large_touch_targets: bool | None = None
    kiosk_auto_dim: bool | None = None
    kiosk_dim_start: str | None = None
    kiosk_dim_end: str | None = None
    kiosk_dim_brightness: int | None = None


class ProfileResponse(ProfileBase):
    """Profile response schema."""

    id: int
    user_id: int
    created_at: str
    updated_at: str | None = None

    model_config = {"from_attributes": True}


class ThemeInfo(BaseModel):
    """Theme information."""

    id: str
    name: str
    description: str
    preview_colors: dict[str, str]

File: frontend/src/lib/api/profiles.ts
/**
 * Profiles API endpoints for themes and user preferences
 */

import { api, type ApiResponse } from './client';

export interface Profile {
	id: number;
	user_id: number;
	theme: string;
	color_scheme: string | null;
	font_size: string;
	contrast: string;
	dashboard_layout: Record<string, unknown> | null;
	favorite_devices: number[] | null;
	default_room: string | null;
	show_weather: boolean;
	show_clock: boolean;
	reduce_motion: boolean;
	high_contrast: boolean;
	large_touch_targets: boolean;
	kiosk_auto_dim: boolean;
	kiosk_dim_start: string | null;
	kiosk_dim_end: string | null;
	kiosk_dim_brightness: number;
	created_at: string;
	updated_at: string | null;
}

export interface ProfileUpdateInput {
	theme?: string;
	color_scheme?: string;
	font_size?: string;
	contrast?: string;
	dashboard_layout?: Record<string, unknown>;
	favorite_devices?: number[];
	default_room?: string;
	show_weather?: boolean;
	show_clock?: boolean;
	reduce_motion?: boolean;
	high_contrast?: boolean;
	large_touch_targets?: boolean;
	kiosk_auto_dim?: boolean;
	kiosk_dim_start?: string;
	kiosk_dim_end?: string;
	kiosk_dim_brightness?: number;
}

export interface ThemeInfo {
	id: string;
	name: string;
	description: string;
	preview_colors: {
		primary: string;
		bg: string;
		surface: string;
	};
}

/**
 * Get available themes
 */
export async function getThemes(): Promise<ApiResponse<ThemeInfo[]>> {
	return api.get<ThemeInfo[]>('/api/v1/profiles/themes');
}

/**
 * Get current user's profile
 */
export async function getMyProfile(): Promise<ApiResponse<Profile>> {
	return api.get<Profile>('/api/v1/profiles/me');
}

/**
 * Update current user's profile
 */
export async function updateMyProfile(updates: ProfileUpdateInput): Promise<ApiResponse<Profile>> {
	return api.patch<Profile>('/api/v1/profiles/me', updates);
}

/**
 * Get a user's profile (admin only or own)
 */
export async function getUserProfile(userId: number): Promise<ApiResponse<Profile>> {
	return api.get<Profile>(`/api/v1/profiles/${userId}`);
}

/**
 * Update a user's profile (admin only or own)
 */
export async function updateUserProfile(
	userId: number,
	updates: ProfileUpdateInput
): Promise<ApiResponse<Profile>> {
	return api.patch<Profile>(`/api/v1/profiles/${userId}`, updates);
}
